File: app/agents/memory.py
"""Forms episodic memories and reinforces existing ones (§9.2, §14)."""

from app.agents.base import Strict, clamp, event_text, fmt_rows, per_message
from app.events.store import Event
from app.services import Services
from app.state.models import LIVE_MEMORY, Proposal

SYSTEM = """You are the memory agent of a persistent-state assistant.
Given one event and the most similar existing memories, decide whether anything is worth
remembering.
Most small talk is not. Prefer reinforcing an existing memory over creating a near-duplicate.
Each new memory is one self-contained sentence stating what happened or what was learned.
Return only the JSON object."""


class NewMemory(Strict):
    summary: str
    importance: float
    confidence: float
    entities: list[str]


class Reinforce(Strict):
    memory_id: str


class MemoryDecision(Strict):
    new_memories: list[NewMemory]
    reinforce: list[Reinforce]


async def run(svc: Services, events: list[Event]) -> list[Proposal]:
    return await per_message(svc, events, _one)


async def _one(svc: Services, e: Event) -> list[Proposal]:
    proposals: list[Proposal] = []
    text = event_text(e)
    (emb,) = await svc.llm.embed([text])
    similar = await svc.repo.similar("memories", emb, 5, LIVE_MEMORY)
    user = (
        f"EVENT:\n[{e['event_id']}] {e['source']}: {text}\n\nSIMILAR MEMORIES:\n"
        f"{fmt_rows(similar, 'summary', 'importance')}"
    )
    out = await svc.llm.complete_json(SYSTEM, user, MemoryDecision)
    known = {m["id"] for m in similar}
    for r in out.reinforce:
        if r.memory_id in known:
            proposals.append(
                Proposal(
                    agent="memory",
                    operation="update_memory",
                    target=r.memory_id,
                    payload={"reinforce": True, "add_source_events": [e["event_id"]]},
                    evidence=[e["event_id"]],
                    confidence=0.8,
                )
            )
    for nm in out.new_memories:
        ents = [x for n in nm.entities if (x := await svc.repo.find_entity(n))]
        imp = clamp(nm.importance)
        proposals.append(
            Proposal(
                agent="memory",
                operation="create_memory",
                payload={
                    "kind": "episodic",
                    "summary": nm.summary,
                    "importance": imp,
                    "confidence": clamp(nm.confidence),
                    "status": "active" if imp >= 0.3 else "candidate",
                    "source_events": [e["event_id"]],
                    "entity_ids": [x["id"] for x in ents],
                },
                evidence=[e["event_id"]],
                confidence=clamp(nm.confidence),
            )
        )
    return proposals

File: src/agents/memory.rs
//! Forms episodic memories and reinforces existing ones (§9.2, §14).

use anyhow::Result;
use serde::{Deserialize, Serialize};
use serde_json::json;

use super::base::{clamp, event_text, fmt_rows, message_events};
use crate::Services;
use crate::llm::MEMORY_DECISION;
use crate::pyfmt::Row;
use crate::state::models::{LIVE_MEMORY, Proposal};

pub const SYSTEM: &str = "You are the memory agent of a persistent-state assistant.
Given one event and the most similar existing memories, decide whether anything is worth
remembering.
Most small talk is not. Prefer reinforcing an existing memory over creating a near-duplicate.
Each new memory is one self-contained sentence stating what happened or what was learned.
Return only the JSON object.";

#[derive(Debug, Default, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct NewMemory {
    pub summary: String,
    pub importance: f64,
    pub confidence: f64,
    pub entities: Vec<String>,
}

#[derive(Debug, Default, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct Reinforce {
    pub memory_id: String,
}

#[derive(Debug, Default, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct MemoryDecision {
    pub new_memories: Vec<NewMemory>,
    pub reinforce: Vec<Reinforce>,
}

pub async fn run(svc: &Services, events: &[Row]) -> Result<Vec<Proposal>> {
    let mut out = Vec::new();
    for e in message_events(events) {
        out.extend(one(svc, &e).await?);
    }
    Ok(out)
}

async fn one(svc: &Services, e: &Row) -> Result<Vec<Proposal>> {
    let mut proposals = Vec::new();
    let text = event_text(e);
    let eid = e["event_id"].as_str().unwrap_or_default().to_string();
    let emb = svc.llm.embed(std::slice::from_ref(&text)).await?.remove(0);
    let similar = svc
        .store
        .lock()
        .unwrap()
        .similar("memories", &emb, 5, Some(LIVE_MEMORY));
    let user = format!(
        "EVENT:\n[{eid}] {}: {text}\n\nSIMILAR MEMORIES:\n{}",
        e["source"].as_str().unwrap_or_default(),
        fmt_rows(&similar, &["summary", "importance"])
    );
    let out: MemoryDecision = svc
        .llm
        .complete_json(SYSTEM, &user, &MEMORY_DECISION)
        .await?;
    let known: Vec<&str> = similar.iter().filter_map(|m| m["id"].as_str()).collect();
    for r in &out.reinforce {
        if known.contains(&r.memory_id.as_str()) {
            proposals.push(
                Proposal::new(
                    "memory",
                    "update_memory",
                    json!({"reinforce": true, "add_source_events": [eid]}),
                )
                .target(&r.memory_id)
                .evidence(vec![eid.clone()])
                .confidence(0.8),
            );
        }
    }
    for nm in &out.new_memories {
        let ents: Vec<String> = {
            let st = svc.store.lock().unwrap();
            nm.entities
                .iter()
                .filter_map(|n| st.state.find_entity(n, None))
                .map(|x| x["id"].as_str().unwrap_or_default().to_string())
                .collect()
        };
        let imp = clamp(nm.importance);
        let payload = json!({
            "kind": "episodic",
            "summary": nm.summary,
            "importance": imp,
            "confidence": clamp(nm.confidence),
            "status": if imp >= 0.3 { "active" } else { "candidate" },
            "source_events": [eid],
            "entity_ids": ents,
        });
        proposals.push(
            Proposal::new("memory", "create_memory", payload)
                .evidence(vec![eid.clone()])
                .confidence(clamp(nm.confidence)),
        );
    }
    Ok(proposals)
}
